src/Confluent.Kafka.Avro/AvroDeserializer.cs

```diff
@@ -0,0 +1,228 @@
+        ///     internally based on configuration properties extracted from the collection
+        ///     passed into the Consumer constructor.
+        /// </remarks>
+        public AvroDeserializer() { }
+
+        /// <summary>
+        ///     Initiliaze a new AvroDeserializer instance.
+        /// </summary>
+        /// <param name="schemaRegisteryClient">
+        ///     An instance of an implementation of ISchemaRegistryClient used for
+        ///     communication with Confluent Schema Registry.
+        /// </param>
+        /// <exception cref="InvalidOperationException">
+        ///	    The generic type <see cref="T"/> is not supported.
+        /// </exception>
+        public AvroDeserializer(ISchemaRegistryClient schemaRegisteryClient)
+        {
+            disposeClientOnDispose = false;
+            SchemaRegistryClient = schemaRegisteryClient;
+        }
+
+        /// <summary>
+        ///     Deserialize an object of type <see cref="T"/> from a byte array.
+        /// </summary>
+        /// <param name="topic">
+        ///     The topic associated with the data (ignored by this deserializer).
+        /// </param>
+        /// <param name="array">
+        ///     A byte array containing the object serialized in the format produced
+        ///     by <see cref="AvroSerializer" />.
+        /// </param>
+        /// <returns>
+        ///     The deserialized <see cref="T"/> value.
+        /// </returns>
+        public T Deserialize(string topic, byte[] array)
+        {
+            // Note: topic is not necessary for deserialization (or knowing if it's a key 
+            // or value) only the schema id is needed.
+
+            using (var stream = new MemoryStream(array))
+            using (var reader = new BinaryReader(stream))
+            {
+                int magicByte = reader.ReadByte();
+                if (magicByte != Constants.MagicByte)
+                {
+                    // may change in the future.
+                    throw new InvalidDataException("magic byte should be 0");
+                }
+                int writerIdBigEndian = reader.ReadInt32();
+
+                DatumReader<T> datumReader = null;
+                // TODO: A r/w lock would be better, but the improvement probably negligible.
+                lock (datumReaderLock)
+                {
+                    datumReaderBySchemaIdBigEndian.TryGetValue(writerIdBigEndian, out datumReader);
+                }
+                if (datumReader == null)
+                {
+                    int writerId = IPAddress.NetworkToHostOrder(writerIdBigEndian);
+                    string writerSchemaJson = SchemaRegistryClient.GetSchemaAsync(writerId).Result;
+                    var writerSchema = Avro.Schema.Parse(writerSchemaJson);
+
+                    datumReader = new SpecificReader<T>(writerSchema, ReaderSchema);
+                    lock (datumReaderLock)
```

Comment:
yes, was aware of that - it's benign though, since the datumReader will just get overwritten and there is no resource leak since SpecificReader is not disposable. Still, I'll put an if statement here to prevent the race condition since that is even simpler than writing a note explaining that it doesn't matter.